cdap-ui/app/cdap/components/Market/index.js

```diff
@@ -49,12 +56,52 @@ export default class Market extends Component {
       }
     });
 
-    MyMarketApi.list().subscribe(this.getCategories, (err) => {
-      console.log('Error', err);
-      MarketAction.setError();
+    if (hasMultipleMarkets()) {
+      this.populateMarketNames();
+      return;
+    }
+
+    this.populateMarketPackages(window.CDAP_CONFIG.marketUrls[0]);
+  }
+
+  populateMarketPackages(marketHost) {
+    MyMarketApi.list({ marketHost }).subscribe(
+      (packages) => {
+        this.getCategories(packages, marketHost);
+      },
+      () => {
+        this.handleMarketConnectionError(marketHost);
+      }
+    );
+  }
+
+  populateMarketNames() {
+    if (!hasMultipleMarkets()) {
+      return;
+    }
+    const markets = window.CDAP_CONFIG.marketUrls;
+
+    // Cache all market names from market metadata.
+    const observables = markets.map((marketHost) =>
+      MyMarketApi.getMetaData({ marketHost }).catch(() => {
+        // Fallback to "Unknown" as the market name. If the user clicks the unknown
+        // market, a connection failure UI would most likly appear.
+        return Observable.of({ marketName: 'Unknown' });
```

Comment:
Can we fallback to the domain name? If there are too many markets without a /metadata api then it becomes hard to track which is a specific market user is looking for.

![Screen Shot 2020-03-09 at 5 40 34 PM](https://user-images.githubusercontent.com/1452845/76268955-185cb280-622d-11ea-8431-9bebf43c6a87.png)

Reply:
Discussed yesterday. I have switched to use market base url for now. let me know if it is acceptable as a short term solution.